app/Models/BeatmapSet.php

```diff
@@ -471,6 +473,177 @@ public function replies($time = null)
+    {
+        if (!self::isValidCoverSize($coverSize)) {
+            return false;
+        }
+
+        $timestamp = 0;
+        if ($this->cover_updated_at) {
+            $timestamp = $this->cover_updated_at->format('U');
+        }
+
+        return $this->storage()->url("/beatmaps/{$this->beatmapset_id}/covers/{$coverSize}.jpg?{$timestamp}");
+    }
+
+    public function storage()
+    {
+        if ($this->_storage === null) {
+            $this->_storage = new StorageWithUrl();
+        }
+
+        return $this->_storage;
+    }
+
+    // todo: generalize method
+    public function oszDownloadURL($noVideo = 1)
+    {
+        $mirrors = config('osu.beatmap_processor.mirrors_to_use');
+        $mirror = BeatmapMirror::find($mirrors[array_rand($mirrors)]);
+
+        $diskFilename = $serveFilename = $this->filename;
+        $time = time();
+        $checksum = md5("{$this->beatmapset_id}{$diskFilename}{$serveFilename}{$time}{$noVideo}{$mirror->secret_key}");
+
+        $url = "{$mirror->base_url}d/{$this->beatmapset_id}?fs=".rawurlencode($serveFilename).'&fd='.rawurlencode($diskFilename)."&ts=$time&cs=$checksum&u=0&nv=$noVideo";
+
+        return $url;
+    }
+
+    public function regenerateCovers()
+    {
+        $tmpBase = sys_get_temp_dir()."/bm/{$this->beatmapset_id}-".time();
+        $workingFolder = "$tmpBase/working";
+        $outputFolder = "$tmpBase/out";
+
+        try {
+            // make our temp folders if they don't exist
+            if (!is_dir($workingFolder)) {
+                mkdir($workingFolder, 0755, true);
+            }
+            if (!is_dir($outputFolder)) {
+                mkdir($outputFolder, 0755, true);
+            }
+
+            // download and extract beatmap
+            $osz = "$tmpBase/osz.zip";
+            $ok = copy($this->oszDownloadURL(), $osz);
+            if (!$ok) {
+                throw new BeatmapProcessorException('Error retrieving beatmap');
+            }
+            $zip = new \ZipArchive;
+            $zip->open($osz);
+            $zip->extractTo($workingFolder);
+            $zip->close();
+
+            // grab the first beatmap (as per old implementation) and scan for background image
+            $beatmap = $this->beatmaps()->first();
+            $beatmapFilename = $beatmap->filename;
+            $bgFilename = self::scanBMForBG("$workingFolder/$beatmapFilename");
+
+            if (!$bgFilename) {
+                $this->update(['cover_updated_at' => $this->freshTimestamp()]);
+
+                return;
+            }
+
+            $bgFile = ci_file_search("{$workingFolder}/{$bgFilename}");
+            if (!$bgFile) {
+                throw new BeatmapProcessorException("Background image missing: {$bgFile}");
+            }
+
+            $processor = new ImageProcessorService($tmpBase);
+
+            // upload original image
+            $this->storage()->put("/beatmaps/{$this->beatmapset_id}/covers/raw.jpg", file_get_contents($bgFile));
```

Comment:
oh apparently I called it storeFile somewhere else. Or maybe in this case storeCover? It's still part of beatmapset since the path contains id anyway. Until you move it to its own class, I guess.